File: backend/app/services/normalizer.py
"""
Query Normalization Service
Standardizes SQL queries by removing obfuscation techniques
"""
import re
from urllib.parse import unquote

class QueryNormalizer:
    def __init__(self):
        self.comment_patterns = [
            (r'--.*$', ''),  # Single line comments
            (r'/\*.*?\*/', ''),  # Multi-line comments
            (r'#.*$', ''),  # MySQL comments
        ]
    
    def normalize(self, query: str) -> str:
        """
        Normalize SQL query by:
        1. URL decoding
        2. Removing comments
        3. Normalizing whitespace
        4. Converting to lowercase
        """
        # URL decode (handle multiple encoding)
        normalized = query
        for _ in range(3):  # Handle triple encoding
            decoded = unquote(normalized)
            if decoded == normalized:
                break
            normalized = decoded
        
        # Remove comments
        for pattern, replacement in self.comment_patterns:
            normalized = re.sub(pattern, replacement, normalized, flags=re.MULTILINE | re.DOTALL)
        
        # Normalize whitespace
        normalized = re.sub(r'\s+', ' ', normalized)
        
        # Strip leading/trailing whitespace
        normalized = normalized.strip()
        
        # Convert to lowercase for consistency
        normalized = normalized.lower()
        
        return normalized
    
    def remove_obfuscation(self, query: str) -> str:
        """Remove common obfuscation techniques"""
        # Remove null bytes
        query = query.replace('\x00', '')
        
        # Remove excessive whitespace variations
        query = re.sub(r'[\t\n\r\f\v]+', ' ', query)
        
        return query

File: customer_integration/backend/services/normalizer.js
/**
 * Query Normalizer Service
 * Standardizes SQL queries by removing obfuscation techniques
 * JavaScript implementation matching Python version
 */

class QueryNormalizer {
    constructor() {
        this.commentPatterns = [
            { pattern: /--.*$/gm, replacement: '' },  // Single line comments
            { pattern: /\/\*.*?\*\//gs, replacement: '' },  // Multi-line comments
            { pattern: /#.*$/gm, replacement: '' },  // MySQL comments
        ];
    }

    /**
     * URL decode helper
     */
    urlDecode(str) {
        try {
            return decodeURIComponent(str);
        } catch (e) {
            return str;
        }
    }

    /**
     * Normalize SQL query by:
     * 1. URL decoding
     * 2. Removing comments
     * 3. Normalizing whitespace
     * 4. Converting to lowercase
     */
    normalize(query) {
        if (!query || typeof query !== 'string') {
            return '';
        }

        // URL decode (handle multiple encoding)
        let normalized = query;
        for (let i = 0; i < 3; i++) {  // Handle triple encoding
            const decoded = this.urlDecode(normalized);
            if (decoded === normalized) {
                break;
            }
            normalized = decoded;
        }

        // Remove comments
        for (const { pattern, replacement } of this.commentPatterns) {
            normalized = normalized.replace(pattern, replacement);
        }

        // Normalize whitespace
        normalized = normalized.replace(/\s+/g, ' ');

        // Strip leading/trailing whitespace
        normalized = normalized.trim();

        // Convert to lowercase for consistency
        normalized = normalized.toLowerCase();

        return normalized;
    }

    /**
     * Remove common obfuscation techniques
     */
    removeObfuscation(query) {
        if (!query || typeof query !== 'string') {
            return '';
        }

        // Remove null bytes
        let cleaned = query.replace(/\x00/g, '');

        // Remove excessive whitespace variations
        cleaned = cleaned.replace(/[\t\n\r\f\v]+/g, ' ');

        return cleaned;
    }
}

module.exports = QueryNormalizer;

